spark/src/main/scala/org/apache/spark/sql/delta/perf/OptimizeMetadataOnlyDeltaQuery.scala

```diff
@@ -60,37 +127,199 @@ trait OptimizeMetadataOnlyDeltaQuery {
     Some(numRecords)
   }
 
-  /** e.g. sql("SELECT COUNT(*) FROM <deta-table>").collect() */
-  object CountStarDeltaTable {
-    def unapply(plan: Aggregate): Option[Long] = plan match {
-      case Aggregate(
-          Nil,
-          Seq(Alias(AggregateExpression(Count(Seq(Literal(1, _))), Complete, false, None, _), _)),
-          PhysicalOperation(_, Nil, DeltaTable(i: TahoeLogFileIndex))) if i.partitionFilters.isEmpty
-        => extractGlobalCount(i)
-      case _ => None
+  val columnStatsSupportedDataTypes: HashSet[DataType] = HashSet(
+    ByteType,
+    ShortType,
+    IntegerType,
+    LongType,
+    FloatType,
+    DoubleType,
+    DateType)
+
+  case class DeltaColumnStat(
+    min: Any,
+    max: Any)
+
+  def extractGlobalColumnStats(tahoeLogFileIndex: TahoeLogFileIndex):
+  CaseInsensitiveMap[DeltaColumnStat] = {
+
+    // TODO Update this to work with DV (https://github.com/delta-io/delta/issues/1485)
+
+    val deltaScanGenerator = getDeltaScanGenerator(tahoeLogFileIndex)
+    val snapshot = deltaScanGenerator.snapshotToScan
+
+    def extractGlobalColumnStatsDeltaLog(snapshot: Snapshot):
+    Map[String, DeltaColumnStat] = {
+
+      val dataColumns = snapshot.statCollectionPhysicalSchema
+        .filter(col => columnStatsSupportedDataTypes.contains(col.dataType))
+
+      // Validate all the files has stats
+      lazy val filesStatsCount = deltaScanGenerator.filesWithStatsForScan(Nil).select(
+        count(when(col("stats.numRecords").isNull, 1)).as("missingNumRecords"),
+        count(when(col("stats.numRecords") > 0, 1)).as("countNonEmptyFiles")).head
+
+      lazy val allRecordsHasStats = filesStatsCount.getAs[Long]("missingNumRecords") == 0
+      lazy val numFiles: Long = filesStatsCount.getAs[Long]("countNonEmptyFiles")
+
+      // DELETE operations creates AddFile records with 0 rows, and no column stats.
+      // We can safely ignore it since there is no data.
+      lazy val files = deltaScanGenerator.filesWithStatsForScan(Nil)
```

Comment:
we also run `deltaScanGenerator.filesWithStatsForScan` for `count(1)` optimization. Should we share the result in case the query has `count(1), min(col1)`.

Reply:
Done